### PR TITLE
Dynamic Rarity

### PR DESCRIPTION
**Pull request by BrownSlaughter**
Tired of seeing Mr mime as ultra rare when you see 100's an hour?
This will set rarity for all pokemon from what you actually scan!

## Description
sets rarity group dependant what you actually see.
Updated pokemon.json

## Motivation and Context
Rarity is area dependent, what one person scans in America is not valid for what someone scans in England so having a 1 size fits all pokemon.json does not work for all. 

Future pokemon will automatically get set a rarity without needing to update pokemon.json

Scaling by rarity will also be more accurate because of this.

## How Has This Been Tested?
looked on map and checked it works like it did before changes but dynamically

Tested on a clean DB everything starts as ultra rare because its never seen them before but quickly changes during initial scan

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by daangroot** [comment]
very nice 👍

**Comment by darkelement1987** [comment]
Nice feature, hope this will make the dev branch :)

**Comment by darkelement1987** [comment]
I'm getting this error:

Traceback (most recent call last):
  File "./runserver.py", line 20, in <module>
    from pogom.app import Pogom
  File "C:\PokemonGo-Map-3.2.0(2)\pogom\app.py", line 13, in <module>
    from pogom.utils import get_args
  File "C:\PokemonGo-Map-3.2.0(2)\pogom\utils.py", line 21, in <module>
    from cHaversine import haversine
ImportError: No module named cHaversine

C:\PokemonGo-Map-3.2.0(2)>

**Comment by BrownSlaughter** [comment]
@darkelement1987 you are not on the latest dev as cHaversine==0.3.0 is in https://github.com/RocketMap/RocketMap/blob/develop/requirements.txt

**Comment by friscoMad** [comment]
@BrownSlaughter have you do some metrics? I am a bit worried about the timing as the function is called for every single pokemon sent to the UI and get_seen() does 2 complex queries and even when it is  cached, the max size seems really small (we have almost 250 pokes now) and it is shared with other callers. So it seems that you can easily reach the limits.
Other than that I really love this to be dynamic.

**Comment by darkelement1987** [comment]
Thanks, i forgot to grab commit #2192. I'll try again later

File: pogom/models.py [hunk 1]
```diff
@@ -27,7 +27,7 @@
 from timeit import default_timer
 
 from . import config
-from .utils import (get_pokemon_name, get_pokemon_rarity, get_pokemon_types,
+from .utils import (get_pokemon_name, i8ln, get_pokemon_types,
```
Comment (Alderon86) [inline pogom/models.py]: im not a fan of that, we need to rework the translation system anyway to properly support translation. but i think for now its ok to have it still around.

**Comment by darkelement1987** [comment]
Got this working since a few days, works great! Please push it into dev!

**Comment by SkOODaT** [comment]
ran this all day yesterday... works perfect so far, i say merge, this is one of the most important features the map needs.....

File: pogom/models.py [hunk 2]
```diff
@@ -328,6 +328,32 @@ def get_appearances_times_by_spawnpoint(cls, pokemon_id,
...
 
+    @staticmethod
+    def get_rarity(pokemon_id):
+        seen = Pokemon.get_seen(0)
+        total = seen['total']
+        found = 0
+        spawn_group = ''
+        for pokemon in seen['pokemon']:
+                if pokemon['pokemon_id'] == pokemon_id:
+                    found = 1
```
Comment (sLoPPydrive) [inline pogom/models.py]: You could actually `break` out of the loop.

File: pogom/models.py [hunk 3]
```diff
@@ -328,6 +328,32 @@ def get_appearances_times_by_spawnpoint(cls, pokemon_id,
 
         return list(itertools.chain(*query))
 
+    @staticmethod
+    def get_rarity(pokemon_id):
+        seen = Pokemon.get_seen(0)
```
Comment (sLoPPydrive) [inline pogom/models.py]: This is kind of an expensive call and although it is being cached somehow you shouldn't really repeat the following steps for each found Pokemon. Instead you should call this **only once** and count all Pokemon occurrences before determining their rarity when you iterate over the found Pokemon.

File: pogom/models.py [hunk 4]
```diff
@@ -328,6 +328,32 @@ def get_appearances_times_by_spawnpoint(cls, pokemon_id,
...
+        spawn_rate = round(100 * pokemon_count / float(total), 4)
+        if spawn_rate > 1:
+            spawn_group = 'Common'
+        elif spawn_rate > 0.5 and spawn_rate < 1:
+            spawn_group = 'Uncommon'
+        elif spawn_rate > 0.03 and spawn_rate < 0.5:
+            spawn_group = 'Rare'
+        elif spawn_rate > 0.01 and spawn_rate < 0.03:
+            spawn_group = 'Very Rare'
+        elif spawn_rate < 0.01:
```
Comment (sLoPPydrive) [inline pogom/models.py]: If you reverse the `elif` chain you could make it easier to read (also you are missing the rare situations where spawn_rate is **equal** to some value:
```
if spawn_rate < 0.01:
  ...
elif spawn_rate < 0.03:
  ...
elif spawn_rate < 0.5:
  ...
elif spawn_rate < 1:
  ...
else:   # anything >= 1
  ...
```

Comment (BrownSlaughter) [reply pogom/models.py]: good spot on the possible equal to possibility, I also like the reveral of the elif chain, I shall make those changes when I get 5 mins

**Comment by fosJoddie** [comment]
getting an error in my console after trying this one:

```
[2017-10-10 14:24:07,776] ERROR in app: Exception on /raw_data [GET]
Traceback (most recent call last):
  File "C:\Python27\lib\site-packages\flask\app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "C:\Python27\lib\site-packages\flask\app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "C:\Python27\lib\site-packages\flask\app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "C:\Python27\lib\site-packages\flask\app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "C:\Python27\lib\site-packages\flask\app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "d:\rm\rm-vanilla\pogom\app.py", line 298, in raw_data
    timestamp=timestamp)
  File "d:\rm\rm-vanilla\pogom\models.py", line 194, in get_active
    p['pokemon_rarity'] = Pokemon.get_rarity(p['pokemon_id'])
  File "d:\rm\rm-vanilla\pogom\models.py", line 346, in get_rarity
    seen = Pokemon.get_seen(0)
  File "C:\Python27\lib\site-packages\cachetools\__init__.py", line 48, in wrapper
    cache[k] = v
  File "C:\Python27\lib\site-packages\cachetools\ttl.py", line 91, in __setitem__
    self.expire(time)
  File "C:\Python27\lib\site-packages\cachetools\ttl.py", line 171, in expire
    cache_delitem(self, curr.key)
  File "C:\Python27\lib\site-packages\cachetools\cache.py", line 64, in __delitem__
    del self.__data[key]
KeyError: (0,)
```

**Comment by fosJoddie** [comment]
After checking out the PR again without some trash in there, it seems to work. There are some issues:

- Initial loading of rarity takes a LONG time (about 10 minutes) with a database of about 6GB from july 1. There needs to be a limit to how far back the calculation looks
- my scanner spans 2 biomes (forest/mountain and trash) .. I see about 10 teddiursas at the same time in the forest for instance.. but beause the trash (pidgey ratata) biome area is larger than the forest/mountain,  I still end up with a rarity rating that is dilluted. I am not sure how to solve such a problem, but geofences could be one way (give it your manager geofences/steps and put stats in categories per area)

**Comment by friscoMad** [comment]
As @fosJoddie @sLoPPydrive  and me reported previously get_seen is really slow, in my main instance with 50Km2 and not sure if 1 or 2 week of data, it needs 3 minutes while it is at it, all other calls to rarity are blocked and the while get_seen(0) (the culprit) is cached it is only for 5 min, so every 5 min the map will block for 3 mins in my setup, which is totally unaceptable.

To solve this I can only think on something like we did in another project that is doing that only once a day and store the results in a file, in this case doing the calcs during startup and then every 12/24 hours and store in memory, we can do that in a separate thread (or some kind of cron task) and by default return everything as common until the first calculation is done.

**Comment by pogo-excalibur** [comment]
Personally, the way I currently handle this is to run an entirely seperate script to modify the values in `pokemon.json`.

My suggestion would be to separate the rarity data out of `pokemon.json` (e.g. `rarity_data.json`) and take this new file outside of version control, and to add a new tool under `Tools/` which modifies this file.

**Comment by pogo-excalibur** [comment]
Heads up: I've -ahem- *borrowed* the `pokemon.json` file from this PR for #2346.

**Comment by neslinesli93** [comment]
what's the status on this pr?

**Comment by BrownSlaughter** [comment]
@neslinesli93 it it currently on hold due to performance issues on super large maps with huge databases. seb is going to be optimising it soon.

**Comment by darkelement1987** [comment]
Any chance it's fixed soon :)?
  

Also anyone with a working fork including #2282 and #2387 ?

**Comment by BrownSlaughter** [comment]
doesnt look like it will happen anytime soon

**Comment by darkelement1987** [comment]
@pogo-excalibur you got a working fork with the 2 PR's?

**Comment by BrownSlaughter** [comment]
@darkelement1987 This is not the place to discuss these things, keep it on topic of development of the feature

**Comment by darkelement1987** [comment]
Then where can i discuss it cause it's also not allowed on Discord lol

**Comment by sebastienvercammen** [comment]
@darkelement1987, @BrownSlaughter is correct, Github is for discussing development and progression of the PR only. Our Discord is for RocketMap user help, not development. We expect people looking to do experimental things with their setups to learn how to do it themselves (and fix it themselves if they break it).

**Comment by darkelement1987** [comment]
conflicts on latest dev branch

image

**Comment by pogo-excalibur** [comment]
@darkelement1987 That would be why this message is at the bottom of the comment chain:
image

**Comment by BrownSlaughter** [comment]
I am running this on the latest and it still works

This PR doesnt use the code`def get_pokemon_rarity(pokemon_id):
    return i8ln(get_pokemon_data(pokemon_id)['rarity'])`

So it looks like you messed up testing the PR

**Comment by BrownSlaughter** [comment]
having to close due to being a dumb ass, will open a new one with changes

**Pull request closed without merging**